### PR TITLE
Update DFI DUSD vault collateral requirement rules

### PR DESCRIPTION
**Pull request by hidiego**
#### What kind of PR is this?:

/kind feature

#### What this PR does / why we need it:
Force 50% DFI collateral on vaults with DUSD loan

### CODE REVIEW

File: src/chainparams.cpp [hunk 1]
```diff
@@ -137,7 +137,7 @@ class CMainParams : public CChainParams {
         consensus.FortCanningCrunchHeight = 1936000; // June 2, 2022.
         consensus.FortCanningSpringHeight = 2033000; // July 6, 2022.
         consensus.FortCanningGreatWorldHeight = 2212000; // Sep 7th, 2022.
-        consensus.FortCanningEpilogueHeight = std::numeric_limits<int>::max();
+        consensus.FortCanningEpilogueHeight = 2218000;
```
Comment (kuegi) [inline src/chainparams.cpp:140]: this is already in the past. Are we planning a rollback?

File: src/masternodes/mn_checks.cpp [hunk 2]
```diff
@@ -3217,17 +3269,28 @@ class CCustomTxApplyVisitor : public CCustomTxVisitor
...
+                    totalCollateralsDUSD += col.nValue;
+                }
+            }
 
-            if (static_cast<int>(height) < consensus.FortCanningHillHeight) {
-                if (totalCollaterals < collateralsLoans.val->totalCollaterals / 2)
-                    return Res::Err("At least 50%% of the collateral must be in DFI when taking a loan.");
+            uint64_t totalCollaterals = totalCollateralsDUSD + totalCollateralsDFI;
+            if (static_cast<int>(height) < consensus.FortCanningHillHeight
+                && totalCollateralsDUSD < collateralsLoans.val->totalCollaterals / 2){
```
[the remark this answers is not in the document]
Comment (mambisi) [reply src/masternodes/mn_checks.cpp:3286]: thanks

File: src/masternodes/mn_checks.cpp [hunk 3]
```diff
@@ -2976,15 +2997,16 @@ class CCustomTxApplyVisitor : public CCustomTxVisitor
...
 
         bool useNextPrice = false, requireLivePrice = false;
         auto collaterals = mnview.GetVaultCollaterals(obj.vaultId);
 
         auto collateralsLoans = mnview.GetLoanCollaterals(obj.vaultId, *collaterals, height, time, useNextPrice, requireLivePrice);
-        if (!collateralsLoans)
-            return std::move(collateralsLoans);
+        if (!collateralsLoans) return std::move(collateralsLoans);
+
+        res = CollateralAddVerifyDUSDUsage(obj.amount.nTokenId, collateralsLoans);
```
[the remark this answers is not in the document]
Comment (prasannavl) [reply src/masternodes/mn_checks.cpp:3008]: Indeed, this is historical, it's the previous implementation, not yet removed. @hidiego could you please rebase the 
new implementation. Try not to have confusing bases, close to fast tracked impl.

**Review by prasannavl — CHANGES_REQUESTED** [review]

File: src/masternodes/mn_checks.cpp [hunk 4]
```diff
@@ -2943,6 +2947,59 @@ class CCustomTxApplyVisitor : public CCustomTxVisitor
...
+        // Heigh checks
+        auto isPostFCH = static_cast<int>(height) >= consensus.FortCanningHillHeight;
+        auto isPreFCH = static_cast<int>(height) < consensus.FortCanningHillHeight;
+        auto isPostFCE = static_cast<int>(height) >= consensus.FortCanningEpilogueHeight;
+        auto isPostFCR = static_cast<int>(height) >= consensus.FortCanningRoadHeight;
+
+        // Condition checks
+        auto isDFIHalfOfAllCollaterall = totalCollateralsDFI < collateralsLoans.totalCollaterals / 2;
+        auto isDFIAndDUSDHalfOfRequiredCollateral = arith_uint256(totalCollaterals)*100 < (arith_uint256(collateralsLoans.totalLoans) * ratio / 2);
+        auto isDFIHalfOfRequiredCollateral = arith_uint256(totalCollateralsDFI)*100 < (arith_uint256(collateralsLoans.totalLoans) * ratio / 2);
```
Comment (kuegi) [inline src/masternodes/mn_checks.cpp:2979]: its called "isHalf" but its true if its NOT half of all collateral

Comment (mambisi) [reply src/masternodes/mn_checks.cpp]: Renamed all variables

**Comment by BenTStark** [comment]
If you want to mint dTokens and DUSD, you need two vaults to get the same effect as before. Wouldn't it be easier from a user's perspective to evaluate Loans and Collaterals in mn_checks.cpp? 
DUSD Loan has to be backed fully by DFI. Then all other loans must be backed by DFI and DUSD.

**Comment by kuegi** [comment]
> If you want to mint dTokens and DUSD, you need two vaults to get the same effect as before. Wouldn't it be easier from a user's perspective to evaluate Loans and Collaterals in mn_checks.cpp? DUSD Loan has to be backed fully by DFI. Then all other loans must be backed by DFI and DUSD.

you do not need 2 vaults. DUSD loan doesn't have to be backed fully by DFI. Its just that as soon as you have DUSD loans in the vault, 50% of your required collateral needs to be DFI. 
Only reason to have two vaults is if you want one 100% DUSD vault where you mint dToken, and one (smaller, otherwise you can combine them) vault with DFI where you mint DUSD. But then I ask the question why you would mint DUSD in the first place?

**Pull request merged**